src/compiler/AST/Common/Type.elm

```diff
@@ -55,7 +63,71 @@ toString type_ =
             "Bool"
 
         List param ->
-            "List " ++ toString param
+            "List " ++ wrapParens param dict
 
         Unit ->
             "()"
+
+
+getTypeVariablesIndex : Type -> Dict Int String
+getTypeVariablesIndex type_ =
+    getTypeVariablesIndexHelp type_ (typeVariablesLetters, Dict.empty)
+    |> Tuple.second
+
+
+getTypeVariablesIndexHelp : Type -> (List String, Dict Int String) -> (List String, Dict Int String)
+getTypeVariablesIndexHelp type_ (freeLetters, dict) =
+    case type_ of
+        Var int ->
+            case freeLetters of
+                [] ->
+                    ([], dict)
+
+                hd :: xs ->
+                    if Dict.member int dict then
+                        (freeLetters, dict)
+                    else
+                        (xs, Dict.insert int hd dict)
+
+        Function t1 t2 ->
+            getTypeVariablesIndexHelp t1 (freeLetters, dict)
+            |> getTypeVariablesIndexHelp t2
+
+        List param ->
+            getTypeVariablesIndexHelp param (freeLetters, dict)
+
+        Int ->
+            (freeLetters, dict)
+
+        Float ->
+            (freeLetters, dict)
+
+        Char ->
+            (freeLetters, dict)
+
+        String ->
+            (freeLetters, dict)
+
+        Bool ->
+            (freeLetters, dict)
+
+        Unit ->
+            (freeLetters, dict)
+
+
+-- Stops at letter `s`, so that outbound variables display `tN`
+typeVariablesLetters =
+    [ "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s" ]
```

Comment:
Interestingly `elm/compiler` does this:
```hs
fromTypeVariableScheme :: Int -> Name
fromTypeVariableScheme scheme =
  runST
  (
    if scheme < 26 then
      do  mba <- newByteArray 1
          writeWord8 mba 0 (0x61 + fromIntegral scheme)
          freeze mba
    else
      do  let (extra, letter) = quotRem scheme 26
          let !size = 1 + getIndexSize extra
          mba <- newByteArray size
          writeWord8 mba 0 (0x61 + fromIntegral letter)
          writeDigitsAtEnd mba size extra
          freeze mba
  )
```
which, I think, roughly translates to `a...z`, then `a1...z1`, then `a2...z2`, etc. into infinity.
**I think what we do is enough** - I mean, if you do so many type variables, something is wrong ... :)

Reply:
yeah, not sure there is anything to gain here... :)

Reply:
I disagree. You're likely to run into problems with this sooner than you realize. Solution to this is very simple. Just have a function `Int -> Name` as elm does and do module of int and lenght of `typeVariablesLetters`. I would also suggest to define this as Array rather than List. I don't think so small simplification is worth the potential trouble.

Reply:
@turboMaCk What's the usecase where you'd run into more than ~15 of those? IMHO it would have to be some monstrous `(a -> b -> c -> d -> ... -> z)` map26 thing. This doesn't concern situations like "there are more than 15 type variables in the whole module / project" - this is just for when printing types.

Although a generic function `Int -> String` would have its beauty - you'd be able to get rid of `freeLetters` entirely and only pass the Dict around :thinking:

Reply:
say person just don't use type annotation in whole program. Also do you already know how you'll be dealing with scopes. Elm don't have "ScopedTypeVariables". What about function with large let block. don't forget `a` on top is not the same as `a` in function in let block and that is not same `a` as a `a` in another function in the let block. You probably don't want to show error message like `expected a but was given a` right?

Reply:
I'm fairly certain errors like `expected a but was given a` are orthogonal to the `Int -> String` stopping at `s` and then continuing with `t19`, `t20`, `t21`, etc. into infinity.

Anyway without an example that with some Elm coding style you can comfortably get into, say, `g` and above, the biggest motivation for "infinite" `Int -> String` function so far is just the simplicity of it.

Reply:
now I see the function https://github.com/elm-in-elm/compiler/pull/43/files#diff-8483228d929c9cb45d74eeaef0d7cc84R46 sorry I missunderstood first comment. So type var name collision can't happen. Though I would still say that Array would be more logical store to store these. Or maybe even Dict given ids are string values.